controllers/metal3.io/host_state_machine.go

```diff
@@ -107,11 +119,50 @@ func (hsm *hostStateMachine) updateHostStateFrom(initialState metal3v1alpha1.Pro
 			}
 		}
 	}
+
+	return true
 }
 
-func (hsm *hostStateMachine) ReconcileState(info *reconcileInfo) actionResult {
+func (hsm *hostStateMachine) checkDelayedHost(info *reconcileInfo) (res bool, actionRes actionResult) {
+
+	// Check if there's a free slot for hosts that have been previously delayed
+	switch info.host.Status.ErrorType {
+	case metal3v1alpha1.ProvisionerBusyError:
+		if !hsm.Provisioner.CanBeProvisioned() {
+			return true, recordActionFailureForBusyProvisioner(info)
+		}
+		// A slot is available, let's cleanup the status and retry
+		info.host.ClearError()
+		return true, actionContinue{}
+	}
+
+	// Make sure the check is re-applied when provisioning an
+	// host not yet tracked by the provisioner
+	switch info.host.Status.Provisioning.State {
+	case metal3v1alpha1.StateInspecting, metal3v1alpha1.StateProvisioning:
+		if hsm.Provisioner.IsProvisioning() {
+			return false, nil
+		}
+		if !hsm.Provisioner.CanBeProvisioned() {
+			return true, recordActionFailureForBusyProvisioner(info)
+		}
+	}
+
+	return false, nil
+}
+
+func (hsm *hostStateMachine) ReconcileState(info *reconcileInfo) (actionRes actionResult) {
 	initialState := hsm.Host.Status.Provisioning.State
-	defer hsm.updateHostStateFrom(initialState, info)
+
+	defer func() {
+		if !hsm.updateHostStateFrom(initialState, info) {
+			actionRes = recordActionFailureForBusyProvisioner(info)
+		}
+	}()
+
+	if delayed, actionRes := hsm.checkDelayedHost(info); delayed {
```

Comment:
We could just check for nil instead of having two return values, like we do for ensureRegistered() on line 172.

Reply:
Done